### ISSUE TITLE
Problem using inversify with Next.js and code elimination

### ISSUE THREAD

**Issue by HeadFox** [inversify/InversifyJS#1352]
We would like to use inversify dependency injection on both server-side and client-side. 
We have some injectable dependencies that are only needed on server-side. 
So we added a condition to only binding it on server-side.
The problem is that we have both implementation (Client and Server) in our client bundle.

We created a file that create a container instance and bind all the injectable dependencies.
```typescript
// container.ts
import { Container } from 'inversify';

import { ConfigRepository } from './ConfigRepository';

import { PrivateConfig } from './PrivateConfig';
import { PublicConfig } from './PublicConfig';

const container: Container = new Container();
const isServer = process.env.NEXT_IS_SERVER;

// Nextjs gives us this value depending on if the code will be executed on server or client during runtime
if (isServer) { 
   container.bind<ConfigRepository>(ConfigRepository).to(PrivateConfig);
} else {
   container.bind<ConfigRepository>(ConfigRepository).to(PublicConfig);
}
export default container;

```
The ConfigRepository look like this 

```typescript
// ConfigRepository.ts
import { injectable } from 'inversify';

import 'reflect-metadata';

@injectable()
export abstract class ConfigRepository {
    abstract get(key: string): boolean | string;
}
```

Our PrivateConfig file look like this (simplified version):
```typescript
// PrivateConfig.ts
import { injectable } from 'inversify';

import { ConfigRepository } from './ConfigRepository';

import 'reflect-metadata';

@injectable()
export class PrivateConfig extends ConfigRepository {
    get(key: string): boolean | string {
        return 'private-token'
    }
}
```

And the PublicConfig is almost the same but with other return data.

Our problem is that the `private-token` value is in our client-side bundle. 

Are we doing this the good way ? 
We don't really know if it's due to our implementation or if it's not possible to conditionally initialize a dependency 

## Your Environment

* Version used: 5.1.1
* Next.js version : 10.2.3
* Webpack version : 5.x

**Comment by magicmatatjahu** [comment]
@HeadFox Hi! As I remember webpack and rollup in the past had problem with treeshaking of TS decorators, so first I would find out if this problem still exists. They added code from every places where you used the decorators.

> Our problem is that the private-token value is in our client-side bundle.

What do you mean here? Do you mean that if you retrieve token by `container.get(ConfigRepository).get()` you get `private-token` or `public-token`? If `private-token` then probably `isServer` is still set to `true` on client-side, so it's a problem with NextJS.

**Comment by jpapini** [comment]
Hello ! I think we figured out the problem.

I have created this code sandbox project:
https://stackblitz.com/edit/nextjs-with-inversify

Take a look at the file `src/ioc.ts`:
We need to use the dynamic import with `require()` with two conditional branches.
Using `typeof window === "undefined"`is required because Next.js hard replace it with a boolean during the build.

Server side code will not be present in the bundle sent to the client browser.

Thanks for your time !
